include/cuco/detail/trie/dynamic_bitset/dynamic_bitset.cuh

```diff
@@ -0,0 +1,359 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/cuda_stream_ref.hpp>
+
+#include <thrust/device_malloc_allocator.h>
+#include <thrust/device_vector.h>
+
+#include <cuda/std/array>
+
+#include <climits>
+#include <cstddef>
+
+namespace cuco {
+namespace experimental {
+namespace detail {
+
+/**
+ * @brief Struct to store ranks of bits at 256-bit intervals
+ */
+struct rank {
+  uint32_t abs_hi_;                    ///< Upper 32 bits of base
+  uint8_t abs_lo_;                     ///< Lower 8 bits of base
+  cuda::std::array<uint8_t, 3> rels_;  ///< Offsets for 64-bit sub-intervals
+
+  /**
+   * @brief Gets base rank of current 256-bit interval
+   *
+   * @return The base rank
+   */
+  __host__ __device__ constexpr uint64_t abs() const noexcept
+  {
+    return (static_cast<uint64_t>(abs_hi_) << 8) | abs_lo_;
+  }
+
+  /**
+   * @brief Sets base rank of current 256-bit interval
+   *
+   * @param abs Base rank
+   */
+  __host__ __device__ constexpr void set_abs(uint64_t abs) noexcept
+  {
+    abs_hi_ = static_cast<uint32_t>(abs >> 8);
+    abs_lo_ = static_cast<uint8_t>(abs);
+  }
+};
+
+/**
+ * @brief Bitvector class with rank and select index structures
+ *
+ * In addition to standard bitvector get/set operations, this class provides
+ * rank and select operation API. It maintains index structures to make both these
+ * new operations close to constant time.
+ * Bitvector construction happens on host, after which the structures are moved to device.
+ * All subsequent read-only operations access device structures only.
+ *
+ * @tparam Allocator Type of allocator used for device storage
+ */
+// TODO: have to use device_malloc_allocator for now otherwise the container cannot grow
+template <class Allocator = thrust::device_malloc_allocator<std::byte>>
+class dynamic_bitset {
+ public:
+  using size_type = std::size_t;  ///< size type to specify bit index
+  using slot_type = uint64_t;     ///< Slot type
```

Comment:
Question: Should this be just `word_type`?

Reply:
Yes, slot_type is a vestige from aow_storage usage

Reply:
63ed55224354aacccde9054fc3ab698e91186b22